Add the ability to create Unix-domain sockets

jamesnvc opened the pull request:
As mentioned on [the Discourse](https://swi-prolog.discourse.group/t/unix-sockets-with-swi/2123/4).

This seems to be working for me now, although I'm not sure why the pldoc comments aren't showing up.

Tested by starting a socket with `nc -lU /tmp/foo.sock` in a terminal, then running

```prolog
?-  unix_socket(Sock),
     unix_connect(Sock, "/tmp/foo.sock"),
     setup_call_cleanup(
         tcp_open_socket(Sock, Stream),
         format(Stream, "hello!~n", []),
          close(Stream)
       ).
```

jamesnvc committed:
Add the ability to create Unix-domain sockets

jamesnvc committed:
Formatting

JanWielemaker commented:
This pull request has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/unix-sockets-with-swi/2123/4

jamesnvc committed:
Throw instead of return false on windows

jamesnvc committed:
Update pldoc

jamesnvc committed:
Use PL_resource_error() instead of a new term + PL_raise_exception

jamesnvc committed:
Update pldoc

jamesnvc committed:
Don't define unix socket predicates on windows

jamesnvc committed:
Rename unix_socket/1 to unix_domain_socket, remove unix_connect

Teaching tcp_connect/2 to handle a unix domain socket, instead of
having a separate connection predicate

jamesnvc commented:
Test usage now looks like:

```prolog
?-  unix_domain_socket(Sock),
     tcp_connect(Sock, "/tmp/foo.sock"),
     setup_call_cleanup(
         tcp_open_socket(Sock, Stream),
         format(Stream, "hello!~n", []),
          close(Stream)
       ).
```

JanWielemaker commented:
Thanks.  Merged after squash.  Added another patch to deal with the server side and which reorganizes things a bit to reuse more code.

jamesnvc commented:
Thanks Jan!
